### ISSUE TITLE
Core Validation: Renderpasses whose attachments point to non-overlapping layers of 2D array images generate validation errors

### ISSUE THREAD

**Issue by DominikWitczakAMD** [KhronosGroup/Vulkan-LoaderAndValidationLayers#2136]
Let X be a renderpass with a total of 6 attachments. X consists of a total of 3 subpasses (S1, S2, S3), each rendering consecutively to a color attachment C1, C2 and C3, and resolving to resolve attachments R1, R2 and R3 (one per subpass).

Trick here is that R1, R2 and R3 all point to the same 2D array image, but - at the same time - it's a unique, nonoverlapping slice. In other words, R1 renders to slice 0, R2 renders to slice 1 and R3 gets slice 2.

Core validation layer doesn't seem to enjoy this. It appears that ``ValidateCmdBufImageLayouts()`` does not recognize the fact that intra-subpass image layout transitions occur at image subresource level, instead of image level. This leads to a weird validation error which claims that:

```
Cannot submit cmd buffer using image (0x5571ac0) [sub-resource: aspectMask 0x1 array layer 0, mip level 0], with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL.
```

FWIW, each resolve attachment uses an ``VkAttachmentDescription::initialLayout`` of ``VK_IMAGE_LAYOUT_UNDEFINED`` and ``VkAttachmentDescription::finalLayout`` of ``VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL``, as well as ``VkSubpassDescription::layout`` value of ``VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL``.

As usual, I have a repro available hidden in my cuff, if an executable would be of help.

**Comment by DominikWitczakAMD** [comment]
I forgot to add: this is for device instances, created with KHR_maintenance1 extension enabled.

**Comment by DominikWitczakAMD** [comment]
Turns out the image actually was a 3D image and the image views were 2D or 2D array, so that validation error actually makes sense. Yay. :)

**Comment by mark-lunarg** [comment]
Heh!  Keep 'em coming.

**Comment by DominikWitczakAMD** [comment]
On another note, I think I've come across a use case where the same error gets reported, even though it shouldn't.

The spec has the following language:

```
When the VkImageSubresourceRange structure is used to select a subset of the slices of a 3D image’s 
mip level in order to create a 2D or 2D array image view of a 3D image created with 
VK_IMAGE_CREATE_2D_ARRAY_COMPATIBLE_BIT, baseArrayLayer and layerCount specify the first slice 
index and the number of slices to include in the created image view. Such an image view can be used as 
a framebuffer attachment that refers only to the specified range of slices of the selected mip level. 

**However, any layout transitions performed on such an attachment view during a render pass instance 
still apply to the entire subresource referenced which includes all the slices of the selected mip level.**

```

Now, let's say we have two 2D array image views, created off the same 3D image:

* Image view A encapsulates slices 0 to 3 inclusive of the parent image
* Image view B encapsulates slices 1 to 2 inclusive of the parent image

Frame 0 uses a renderpass, whose (only) subpass renders to A. InitialLayout = UNDEFINED, finalLayout = layout = COLOR_ATT_OPTIMAL. The renderpass is followed by an image barrier which transitions the whole image (ie all slices) from COLOR_ATT_OPTIMAL to TRANSFER_SRC_OPTIMAL.

Frame 1 uses a renderpass, whose (only) subpass renders to B. InitialLayout = UNDEFINED, finalLayout = layout = COLOR_ATT_OPTIMAL. The renderpass is again followed by the same barrier as above.

I'm seeing the core validation layer cry the following out in terror when submitting cmd buffers related to frame 1:

``
Cannot submit cmd buffer using image (0x13f408c0)  [sub-resource: aspectMask 0x1 array layer 0, mip level 0], with layout VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL when first use is VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL.
``

Drawing 2x frame 0 in a row does not lead to the error above.
Drawing 2x frame 1 in a row gives:

``
Cannot submit cmd buffer using image (0x4e37c50) [sub-resource: aspectMask 0x1 array layer 0, mip 
level 0], with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is 
VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL.
``

My understanding of the language above marked with ** is that the initial->layout transition of a 2D array attachment whose parent image is a 3D image transfers all slices of the 3D image at once, meaning the errors do not make much sense.

**Comment by chrisforbes** [comment]
@DominikWitczakAMD are there corresponding CTS cases for these unusual usages?

**Comment by DominikWitczakAMD** [comment]
@chrisforbes : As strongly as I'd like to refrain from saying that, I doubt that :-( This one came afloat during our internal testing. I can share a repro if it would be of help.

**Comment by daveh-lunarg** [comment]
I don't grok this completely but I think it's (at least partially) fixed by PR #2408.  I'm closing it based on that  general intuition alone.  @DominikWitczakAMD , if it reappears please re-open this issue or create a new one and assign to me.

**Comment by daveh-lunarg** [comment]
Jumped the gun closing this one - found a problem with recording 3D image layout via a 2D imageview.  Re-opening and will close with a PR.
